File: comscore/src/main/java/com/bitmovin/player/integration/comscore/ComScoreAnalytics.java
package com.bitmovin.player.integration.comscore;

import android.content.Context;
import android.util.Log;

import com.bitmovin.player.BitmovinPlayer;
import com.comscore.Analytics;
import com.comscore.PublisherConfiguration;

import java.util.HashMap;
import java.util.Map;

public class ComScoreAnalytics {
    private static final String TAG = "ComScoreAnalytics";
    private static boolean started;
    private static ComScoreConfiguration configuration;

    /**
     * Starts the ComScoreAnalytics app level tracking
     *
     * @param configuration - ComScore configuration with your app publisher id, publisher secret, and application name
     * @param context       - Application Context
     */
    public static synchronized void start(ComScoreConfiguration configuration, Context context) {
        if (!started) {
            ComScoreAnalytics.configuration = configuration;

            PublisherConfiguration.Builder builder = new PublisherConfiguration.Builder()
                    .publisherId(configuration.getPublisherId())
                    .publisherSecret(configuration.getPublisherSecret())
                    .secureTransmission(configuration.isSecureTransmissionEnabled())
                    .applicationName(configuration.getApplicationName());

            ComScoreUserConsent userConsent = configuration.getUserConsent();
            // Only populate label if user consent is known
            if (userConsent != ComScoreUserConsent.UNKNOWN) {
                Map<String, String> labels = new HashMap<>();
                labels.put("cs_ucfr", userConsent.getValue());
                builder.persistentLabels(labels);
            }

            PublisherConfiguration myPublisherConfig = builder.build();
            Analytics.getConfiguration().addClient(myPublisherConfig);

            Analytics.start(context);
            started = true;
        }
    }

    /**
     * Sets the user consent to granted. Use after the ComScoreAnalytics object has been started
     */
    public static synchronized void userConsentGranted() {
        if (started) {
            PublisherConfiguration publisherConfig = Analytics.getConfiguration().getPublisherConfiguration(ComScoreAnalytics.configuration.getPublisherId());
            publisherConfig.setPersistentLabel("cs_ucfr", ComScoreUserConsent.GRANTED.getValue());
            Analytics.notifyHiddenEvent();
        }
    }

    /**
     * Sets the user consent to denied. Use after the ComScoreAnalytics object has been started
     */
    public static synchronized void userConsentDenied() {
        if (started) {
            PublisherConfiguration publisherConfig = Analytics.getConfiguration().getPublisherConfiguration(ComScoreAnalytics.configuration.getPublisherId());
            publisherConfig.setPersistentLabel("cs_ucfr", ComScoreUserConsent.DENIED.getValue());
            Analytics.notifyHiddenEvent();
        }
    }

    /**
     * Creates ComScoreStreamingAnalytics object that is attached to your bitmovin player
     *
     * @param bitmovinPlayer - the player to report on
     * @param metadata       - ComScoreMetadata associated with the current loaded source
     * @return ComScoreStreamingAnalytics object
     */
    public static synchronized ComScoreStreamingAnalytics createComScoreStreamingAnalytics(BitmovinPlayer bitmovinPlayer, ComScoreMetadata metadata) {
        if (started) {
            return new ComScoreStreamingAnalytics(bitmovinPlayer, ComScoreAnalytics.configuration, metadata);
        } else {
            Log.e(TAG, "ComScoreStreamingAnalytics was not created. Must call start() first");
            throw new ComScoreAnalyticsException("ComScoreStreamingAnalytics was not created. Must call start() first");
        }
    }

    /**
     * Returns true if the ComScoreAnalytics object has been started. You must do this prior to creating ComScoreStreamingAnalytics
     *
     * @return
     */
    public static boolean isStarted() {
        return started;
    }
}

File: comscore/src/main/java/com/bitmovin/player/integration/comscore/ComScoreAnalytics.kt
package com.bitmovin.player.integration.comscore

import android.content.Context

import com.bitmovin.player.BitmovinPlayer
import com.bitmovin.player.integration.comscore.util.notifyHiddenEvent
import com.bitmovin.player.integration.comscore.util.notifyHiddenEvents
import com.comscore.Analytics
import com.comscore.PublisherConfiguration

object ComScoreAnalytics {

    var isStarted: Boolean = false
    private lateinit var configuration: ComScoreConfiguration

    /**
     * Start [ComScoreAnalytics] app level tracking
     *
     * @param configuration - [ComScoreConfiguration] with app publisher id, publisher secret, and application name
     * @param context       - Application context
     */
    @Synchronized
    fun start(configuration: ComScoreConfiguration, context: Context) {
        if (!isStarted) {
            this.configuration = configuration

            val publisherConfig = PublisherConfiguration.Builder().apply {
                publisherId(configuration.publisherId)
                secureTransmission(configuration.secureTransmission)

                // Only set user consent value if not equal to UNKNOWN
                if (configuration.userConsent != ComScoreUserConsent.UNKNOWN) {
                    persistentLabels(mapOf("cs_ucfr" to configuration.userConsent.value))
                }

            }.build()

            with(Analytics.getConfiguration()) {
                addClient(publisherConfig)
                setApplicationName(configuration.applicationName)

                if (configuration.childDirectedAppMode) {
                    enableChildDirectedApplicationMode()
                }
            }
            Analytics.start(context)

            isStarted = true
        }
    }

    /**
     * Set persistent labels on the ComScore [PublisherConfiguration]
     *
     * @param labels - the labels to set
     */
    @Synchronized
    fun setPersistentLabels(labels: Map<String, String>) {
        if (isStarted) {
            notifyHiddenEvents(configuration.publisherId, labels)
            BitLog.d("ComScore persistent labels set: ${labels.map { "${it.key}:${it.value}" }}")
        }
    }

    /**
     * Set a persistent label on the ComScore [PublisherConfiguration]
     *
     * @param label - the label name
     * @param value - the label value
     */
    @Synchronized
    fun setPersistentLabel(label: String, value: String) {
        if (isStarted) {
            notifyHiddenEvent(configuration.publisherId, label, value)
            BitLog.d("ComScore persistent label set: [$label:$value]")
        }
    }

    /**
     * Create [ComScoreStreamingAnalytics] object
     *
     * @param bitmovinPlayer - the [BitmovinPlayer] to report on
     * @param metadata       - [ComScoreMetadata] associated with the current loaded source
     * @return [ComScoreStreamingAnalytics] object
     */
    @Synchronized
    fun createComScoreStreamingAnalytics(bitmovinPlayer: BitmovinPlayer, metadata: ComScoreMetadata): ComScoreStreamingAnalytics {
        if (isStarted) {
            return ComScoreStreamingAnalytics(bitmovinPlayer, configuration, metadata)
        } else {
            BitLog.e("ComScoreStreamingAnalytics was not created. Must call start() first")
            throw ComScoreAnalyticsException("ComScoreStreamingAnalytics was not created. Must call start() first")
        }
    }
}
